app/src/main/java/org/mozilla/tv/firefox/focus/FocusRepo.kt

```diff
@@ -0,0 +1,332 @@
+
+import android.view.View
+import android.view.ViewTreeObserver
+import androidx.annotation.VisibleForTesting
+import io.reactivex.Observable
+import io.reactivex.rxkotlin.Observables
+import io.reactivex.subjects.BehaviorSubject
+import mozilla.components.browser.engine.system.NestedWebView
+import org.mozilla.tv.firefox.R
+import org.mozilla.tv.firefox.ScreenController
+import org.mozilla.tv.firefox.ScreenControllerStateMachine.ActiveScreen
+import org.mozilla.tv.firefox.pinnedtile.PinnedTileRepo
+import org.mozilla.tv.firefox.pocket.PocketVideoRepo
+import org.mozilla.tv.firefox.session.SessionRepo
+import org.mozilla.tv.firefox.utils.URLs
+
+private const val NESTED_WEB_VIEW_ID = 2147483646 // Int.MAX_VALUE - 1
+
+class FocusRepo(
+    screenController: ScreenController,
+    sessionRepo: SessionRepo,
+    pinnedTileRepo: PinnedTileRepo,
+    pocketRepo: PocketVideoRepo
+) : ViewTreeObserver.OnGlobalFocusChangeListener {
+
+    val defaultFocusMap: HashMap<ActiveScreen, Int> = HashMap()
+
+    data class State(
+        val focusNode: FocusNode,
+        val focused: Boolean = true
+    )
+
+    /**
+     * FocusNode describes quasi-directional focusable paths given viewId
+     */
+    data class FocusNode(
+        val viewId: Int,
+        val nextFocusUpId: Int? = null,
+        val nextFocusDownId: Int? = null,
+        val nextFocusLeftId: Int? = null,
+        val nextFocusRightId: Int? = null
+    ) {
+        fun updateViewNodeTree(view: View) {
+            assert(view.id == viewId)
+
+            nextFocusUpId?.let { view.nextFocusUpId = it }
+            nextFocusDownId?.let { view.nextFocusDownId = it }
+            nextFocusLeftId?.let { view.nextFocusLeftId = it }
+            nextFocusRightId?.let { view.nextFocusRightId = it }
+        }
+    }
+
+    init {
+        initializeDefaultFocus()
+    }
+
+    private fun initializeDefaultFocus() {
+        defaultFocusMap[ActiveScreen.NAVIGATION_OVERLAY] = R.id.navUrlInput
+        defaultFocusMap[ActiveScreen.WEB_RENDER] = R.id.engineView
+        defaultFocusMap[ActiveScreen.POCKET] = R.id.videoFeed
+    }
+
+    // TODO: potential for telemetry?
+    private val _state: BehaviorSubject<State> = BehaviorSubject.createDefault(
+            State(FocusNode(R.id.navUrlInput), focused = true)
+    )
+
+    private val _focusUpdate = Observables.combineLatest(
+            _state,
+            screenController.currentActiveScreen,
+            sessionRepo.state,
+            pinnedTileRepo.isEmpty,
+            pocketRepo.feedState) { state, activeScreen, sessionState, pinnedTilesIsEmpty, pocketState ->
+        dispatchFocusUpdates(state.focusNode, activeScreen, sessionState, pinnedTilesIsEmpty, pocketState)
+    }
+
+    val focusUpdate: Observable<State> = _focusUpdate
+            .distinctUntilChanged()
+            .hide()
+
+    override fun onGlobalFocusChanged(oldFocus: View?, newFocus: View?) {
+        fun <T> BehaviorSubject<T>.onNextIfNew(value: T) {
```

Comment:
Most of `onNextIfNew` is the same thing as just putting `distinctUntilChanged` on the `Observable`.  We can add that operator, get rid of most of this, and then just have (`if (newState.focusNode.viewId != NESTED_WEB_VIEW_ID) this.onNext(value)`.